src/nvim/msgpack_rpc/channel.c

```diff
@@ -246,25 +259,31 @@ static void parse_msgpack(Channel *channel)
   while ((result = msgpack_unpacker_next(channel->rpc.unpacker, &unpacked)) ==
          MSGPACK_UNPACK_SUCCESS) {
     bool is_response = is_rpc_response(&unpacked.data);
-    log_client_msg(channel->id, !is_response, unpacked.data);
+    log_client_msg(DEBUG_LOG_LEVEL, channel->id, !is_response,
+                   unpacked.data);
 
     if (is_response) {
-      if (is_valid_rpc_response(&unpacked.data, channel)) {
+      uint64_t response_id;
+      if (is_valid_rpc_response(&unpacked.data, channel, &response_id)) {
         complete_call(&unpacked.data, channel);
       } else {
         char buf[256];
         snprintf(buf, sizeof(buf),
-                 "ch %" PRIu64 " returned a response with an unknown request "
-                 "id. Ensure the client is properly synchronized",
-                 channel->id);
+                 "ch %" PRIu64 " got response for unknown request-id=%" PRIu64
+                 ". Ensure the client is synchronized",
+                 channel->id,
+                 response_id);
         call_set_error(channel, buf, ERROR_LOG_LEVEL);
       }
       msgpack_unpacked_destroy(&unpacked);
       // Bail out from this event loop iteration
-      return;
+      return false;
     }
 
-    handle_request(channel, &unpacked.data);
+    if (!handle_request(channel, &unpacked.data)) {
+      // Bail out from this event loop iteration
+      return false;
+    }
```

Comment:
We already bail out early on line 280. Without this change, `msgpack_unpacker_next` keeps iterating through every byte, then `handle_request` fails again and again.